### PR TITLE
fix(#2698): honor isolated HERMES_HOME profile mode

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- The first pass hid other profiles and blocked direct mutations, but explicit home resolution, same-name default switching, scheduled cron jobs, and aggregate/import side paths could still fall back to the shared root profile.
- The `#4449` default state-dir contract is part of the same boundary: if isolated deployments still default state or cwd back to shared locations, the runtime remains cross-profile even when the UI says it is pinned.
- The fix makes the startup `HERMES_HOME=.../profiles/<name>` path authoritative for runtime home resolution, aggregate reads, cron execution, startup state-dir defaults, and the remaining profile-management affordances, while leaving normal multi-profile installs unchanged.

## What Changed

- `api/profiles.py`: derive isolated mode from the startup `HERMES_HOME` snapshot, pin active profile/home resolution to that startup path, clamp all explicit profile-home lookups to the isolated home, keep `switch_profile("default")` and scheduled cron jobs on the pinned home, and avoid the isolated fallback recursion when upstream profile helpers are unavailable.
- `api/routes.py`: expose `single_profile_mode`, return 403 for isolated-mode create/switch/delete attempts, ignore `all_profiles=1` on sessions/projects/search while isolated mode is active, and reject `all_profiles` CLI imports in isolated mode.
- `api/config.py` and `bootstrap.py`: fold the `#4449` contract into this branch by defaulting `HERMES_WEBUI_STATE_DIR` to `$HERMES_HOME/webui`, and add `HERMES_WEBUI_SERVER_CWD` so read-only agent installs can redirect fallback relative writes to a writable workspace, including the Windows foreground launch path.
- `static/panels.js`: hide create, activate, delete, manage, and dropdown affordances in single-profile mode.
- `tests/...`: cover isolated startup pinning, `profiles/default`, explicit foreign lookup clamping, cron completion publishing, `all_profiles` import rejection, Windows foreground cwd propagation, the state-dir contract, and the profile-list cache isolation used by the visibility tests.

## Why It Matters

Isolated deployments now keep the reported profile, runtime home, state dir, cron execution, and aggregate read paths on the same pinned profile instead of letting a few fallback paths reopen the shared root profile.

## Verification

- `pytest tests/test_issue2698_isolated_hermes_home.py tests/test_profile_env_isolation.py tests/test_issue1611_session_profile_filtering.py tests/test_bootstrap_foreground.py tests/test_issue4449_state_dir_contract.py tests/test_issue4067_import_cli_cross_profile_guard.py tests/test_scheduled_jobs_profile_isolation.py tests/test_issue3623_profile_visibility.py tests/test_settings_navigation_and_detail_refresh.py -v --timeout=60`
- `npx eslint --no-config-lookup -c eslint.runtime-guard.config.mjs "static/**/*.js"`
- CI context: `pytest tests/ -v --timeout=60`

## Risks / Follow-ups

- Literal-path detection still depends on `HERMES_HOME` being configured as `.../profiles/<name>`; symlinked homes warn instead of silently opting out.
- This keeps profile isolation consistent for the current WebUI surfaces; launcher-level workspace policy is still whatever the operator configures.

## Upstream

Refs #2698. Folds the `#4449` default state-dir contract into the same isolation fix.

## Model Used

GPT-5.5 via Codex CLI

### CODE REVIEW

File: api/routes.py [hunk 1]
```diff
@@ -9869,6 +9879,8 @@ def handle_post(handler, parsed) -> bool:
                 model_provider=model_provider,
             )
             return j(handler, {"ok": True, "profile": result})
+        except PermissionError as e:
+            return bad(handler, str(e), 403)
         except (ValueError, FileExistsError, RuntimeError) as e:
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/routes.py:9892]: Good catch on the consistency gap. The current isolation error message contains no path info, but using \_sanitize_error(e)\ aligns with the project's defensive hygiene pattern. Will address in the next push.

File: api/profiles.py [hunk 2]
```diff
@@ -978,17 +1005,29 @@ def switch_profile(name: str, *, process_wide: bool = True) -> dict:
...
+
     Args:
         name: Profile name to switch to.
         process_wide: If True (default), updates the process-global
             _active_profile.  Set to False for per-client switches from the
             WebUI where the profile is managed via cookie + thread-local (#798).
 
     Returns: {'profiles': [...], 'active': name}
-    Raises ValueError if profile doesn't exist or agent is busy.
+    Raises PermissionError if profile doesn't exist or agent is busy.
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/profiles.py]: Fixed — docstring now accurately states ValueError for missing profile, RuntimeError for busy agent, PermissionError only for isolated-mode cross-profile switch.

File: api/profiles.py [hunk 3]
```diff
@@ -1292,6 +1334,50 @@ def list_profiles_api() -> list:
...
+        try:
+            from hermes_cli.profiles import list_profiles
+            infos = list_profiles()
+            # Find the current profile and return only that one
+            for p in infos:
+                if p.name == active:
+                    enabled_count, total_count = _get_profile_skills_stats(p.path)
+                    return [{
+                        'name': p.name,
+                        'path': str(p.path),
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/profiles.py:1417]: Fixed in 8dc261af. Broadened the except clause from \ImportError\ to \(ImportError, OSError, PermissionError)\ so runtime errors on corrupt profile directories hit the fallback path instead of propagating as 500s.

File: api/profiles.py [hunk 4]
```diff
@@ -116,6 +116,33 @@ def _unwrap_profile_home_to_base(home: Path) -> Path:
...
+    p = Path(hermes_home).expanduser()
+    # Check if this path looks like ~/.hermes/profiles/<name>
+    # i.e., parent dir is named 'profiles' and grandparent exists
+    if p.parent.name == 'profiles' and p.parent.parent.exists():
+        return True
+    return False
+
+
 def _resolve_base_hermes_home() -> Path:
     """Return the BASE ~/.hermes directory — the root that contains profiles/.
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/profiles.py:176]: Fixed in 1ed35a1f. Added _INITIAL_HERMES_HOME snapshot at module import time (before init_profile_state can overwrite the env var). _is_isolated_profile_mode() now reads the snapshot instead of os.environ, so isolation detection survives the startup sequence.

File: api/profiles.py [hunk 5]
```diff
@@ -1292,6 +1347,50 @@ def list_profiles_api() -> list:
...
+            'is_active': True,
+            'gateway_running': False,
+            'model': None,
+            'provider': None,
+            'has_env': (hermes_home / '.env').exists(),
+            'visible': _profile_visible_from_meta(hermes_home),
+            'skill_count': enabled_count,
+            'enabled_skills': enabled_count,
+            'total_skills': total_count,
+        }]
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/profiles.py:1446]: Fixed in 71ffcf3e. Added `_isolated_profile_name()` helper that derives the name from `Path(_INITIAL_HERMES_HOME).name` instead of relying on `get_active_profile_name()`, which returns the process-level `_active_profile` default ("default") before any profile switch has occurred.

File: tests/test_issue2698_isolated_hermes_home.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+import tempfile
+from pathlib import Path
+from unittest import mock
+
+import pytest
+
+from api.profiles import (
+    _is_isolated_profile_mode,
+    list_profiles_api,
+    create_profile_api,
```
[the remark this answers is not in the document]
Comment (rodboev) [reply tests/test_issue2698_isolated_hermes_home.py:26]: Fixed in 9790255d. Added `test_switch_to_different_profile_rejected` (cross-profile switch raises PermissionError) and `test_switch_to_same_profile_idempotent` (same-name switch passes through the isolation guard without raising).

File: api/profiles.py [hunk 7]
```diff
@@ -1292,6 +1352,50 @@ def list_profiles_api() -> list:
...
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/profiles.py:1446]: Fixed in 0072ef6d. Replaced `get_active_hermes_home()` with `Path(_INITIAL_HERMES_HOME).expanduser()` in the isolated-mode fallback path, so the profile home, has_env, and visibility all resolve against the actual isolated directory.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Thanks @rodboev — the shape here is right and the server-side enforcement is genuinely well done. I ran the full gate (Codex + Opus, both adversarial) and verified the good parts: the guard surface is **complete** — `switch`/`create`/`delete` are the only profile-mutation endpoints (no rename/clone/reset/import routes; "clone" is a `create` param), all three raise `PermissionError`→403 as the **first line** of the API function, so a manual `POST /api/profile/switch {"name":"other"}` is rejected regardless of the hidden UI. The `panels.js` changes are cosmetic-only. The import-time `_INITIAL_HERMES_HOME` snapshot timing is correct. Good work on making this a real server boundary.

But there's one issue that blocks shipping, because it makes the feature give a **false assurance**:

### (must-fix) Isolation *detection* and the *active data-home* can diverge — the UI says "pinned to webui" while the runtime serves a different profile's data

`_is_isolated_profile_mode()` / `_isolated_profile_name()` are derived from `_INITIAL_HERMES_HOME` (the path), but the **active profile and its data home are resolved independently** from the base `active_profile` file (`_read_active_profile_file` → `get_active_hermes_home`). In the standard "set `HERMES_HOME=~/.hermes/profiles/webui` and launch" deployment (no matching `active_profile` file written), the two diverge:
- `list_profiles_api` + the `single_profile_mode` flag report the deployment as pinned to `webui`, and the switch/create/delete guards pin to `webui`,
- …but the runtime actually serves the **default/root** profile's sessions/data (because the active-profile resolution still consults the base `active_profile`).

So the operator is told "this instance is confined to the `webli` profile" while it's serving another profile's data — which directly undercuts the PR's value prop and is worse than no isolation claim at all. (It is NOT a leak of other *named* profiles, and the unwrap behavior pre-dates this PR — but layering an isolation guarantee on top of a runtime that may serve a different profile is the false-assurance part.)

**Fix:** make the runtime actually pin to the isolated profile so detection and data-home agree.
1. `api/profiles.py` `get_active_profile_name()` (~:324): when `_is_isolated_profile_mode()` is true, return `_isolated_profile_name()` **before** consulting the thread-local cookie / `_active_profile` / the base `active_profile` file.
2. `init_profile_state()` (~:1005): in isolated mode, initialize `HERMES_HOME`/the active data-home to the isolated profile path, not the shared `active_profile` target.
3. Add a test asserting that in isolated mode (`HERMES_HOME=…/profiles/X` with NO base `active_profile` written, or one pointing elsewhere), `get_active_profile_name()` == `X` AND the served session data resolves to `X`'s home — i.e. detection and data-home cannot diverge.

### (should-fix, low) Symlinked `HERMES_HOME` → silent fail-open
Detection uses the literal `_INITIAL_HERMES_HOME` string with `expanduser()` but no `resolve()`, so if `HERMES_HOME` is a symlink whose own path isn't shaped `…/profiles/<name>`, isolation is silently OFF (guards disabled — fail-open). Don't blindly add `resolve()` (that creates the opposite false-negative when `<name>` itself is a symlink); instead document the contract (HERMES_HOME must be the literal `…/profiles/<name>` path) and/or log a one-line warning when a symlinked home can't be classified, so the toggle never disables itself silently.

Once detection and the active data-home are guaranteed to agree (fix #1), I'm happy to re-gate and ship — the enforcement layer is already solid, this is about making the pinned-profile claim actually true at runtime.

**Comment by rodboev** [comment]
Thanks, this is the right blocker. You're right that the current branch pins the detection and enforcement layer, but not the runtime data-home resolution: `list_profiles_api()` and `single_profile_mode` can present `webui`, while `get_active_profile_name()` and startup still let the base `active_profile` path drive `HERMES_HOME`.

1. `98773cc0` makes isolated mode authoritative in `api/profiles.py`: `get_active_profile_name()` now short-circuits to `_isolated_profile_name()` before TLS or `_active_profile`, and `init_profile_state()` now initializes both `_active_profile` and `HERMES_HOME` from the isolated startup path itself instead of the base `active_profile` file.
2. `cc01ba15` closes the remaining `profiles/default` edge case by making `get_active_hermes_home()` use the literal isolated startup path while isolation is active, so later reads cannot unwrap back to the shared home just because the isolated directory name happens to be `default`.
3. I tightened the regression coverage around startup, not just the detector helpers: isolated `HERMES_HOME` with no base `active_profile`, with a conflicting one, and with the literal `profiles/default` edge case, all asserting that both the active name and the resolved home stay on the isolated directory. I also reran `tests/test_profile_env_isolation.py` because `init_profile_state()` changed.
4. On the symlink point, I kept the detector literal instead of adding a blanket `resolve()`, and added a one-time warning that states the contract when a symlinked `HERMES_HOME` cannot be classified for isolated mode, so it no longer fails open silently.

**Comment by nesquena-hermes** [comment]
Re-gated the new commits (`98773cc0` "pin isolated runtime home to configured profile" and `cc01ba15` "keep isolated default-named profiles pinned") against the three must-fix items from the prior review. All three land, and the divergence I flagged — detection saying "pinned to `webui`" while the runtime serves another profile's data — is now closed at the source.

## What I verified

**Must-fix #1 — `get_active_profile_name()` short-circuits before TLS/global.** `api/profiles.py:334` now resolves isolation first:

```python
def get_active_profile_name() -> str:
    if _is_isolated_profile_mode():
        return _isolated_profile_name()
    tls_name = getattr(_tls, 'profile', None)
    if tls_name is not None:
        return tls_name
    return _active_profile
```

`test_get_active_profile_name_ignores_tls_and_global_in_isolated_mode` (tests/test_issue2698_isolated_hermes_home.py:198) pins `_active_profile="other_profile"` and still asserts `== "user1"`, so neither the cookie TLS nor the process global can leak through.

**Must-fix #2 — startup + data-home pin to the isolated path.** `init_profile_state()` (api/profiles.py:1020) now branches on isolation and ignores the base `active_profile` file entirely:

```python
if _is_isolated_profile_mode():
    _active_profile = _isolated_profile_name()
    home = Path(_INITIAL_HERMES_HOME).expanduser()
else:
    _active_profile = _read_active_profile_file()
    home = get_active_hermes_home()
```

and `get_active_hermes_home()` (api/profiles.py:385) returns `Path(_INITIAL_HERMES_HOME).expanduser()` while isolation is active, so a later read cannot unwrap back to the shared home. This is the part that previously let the operator be told "confined to `webui`" while serving root data.

**Must-fix #3 — a test that detection and data-home cannot diverge.** `test_init_profile_state_pins_runtime_home_to_isolated_profile` (tests/...:215) is parametrized over `[None, "user2"]` for the base `active_profile` contents (missing *and* conflicting) and asserts all of `_active_profile == "user1"`, `get_active_profile_name() == "user1"`, `get_active_hermes_home() == temp_single_profile`, and `os.environ["HERMES_HOME"]` agree. `test_get_active_hermes_home_keeps_profiles_default_pinned` (:259) covers the nasty `profiles/default` edge where the isolated dir name collides with the root alias.

## Two confirmations on the periphery

- The greptile "no Active badge on fresh deploy" gap is transitively closed: the `GET /api/profiles` handler (api/routes.py:8366-8378) feeds `active` from `get_active_profile_name()`, which now returns the isolated name, so `data.active === user1` and the card renders the badge. No separate routes change was needed.
- The symlink should-fix is handled the way I asked — `_is_isolated_profile_mode()` (api/profiles.py:128) keeps the literal check and emits a one-shot warning when a symlinked `HERMES_HOME` can't be classified, instead of a blanket `resolve()` (which would have created the opposite false-negative when `<name>` itself is a symlink).

## Cross-layer note (not a blocker)
The agent's canonical layout matches the detector's `p.parent.name == 'profiles'` assumption — `hermes_cli/profiles.py:251-252` documents profiles living under `HERMES_HOME/profiles/<name>`, and `_get_active_profile_path()` (profiles.py:269) reads the sticky `active_profile` from the *default root* home. That shared root file is exactly what isolated mode now bypasses, which is the correct call: in an isolated deployment the operator deliberately confined the instance to the subdir, and the agent writes that root file as process-global state the WebUI shouldn't follow.

Enforcement layer was already solid; with detection and the runtime data-home now provably in agreement, this is good to ship from my side. Nice work threading the `profiles/default` edge.

**Comment by nesquena-hermes** [comment]
Thanks @rodboev — this is close, and the core design (snapshot `_INITIAL_HERMES_HOME` + pin from the path tail) is the right shape. I deep-reviewed it through a full Codex + Opus + suite gate. The **common case — a named profile like `HERMES_HOME=~/.hermes/profiles/work`** — is correct, fully scoped, leak-free, and backward-compatible (a default install with `HERMES_HOME` unset or `==` base home is provably inert). 

I went ahead and fixed three edge bugs the first gate pass found, and **pushed them to `warmup/4454-isolated-profile-3fixes`** so you can cherry-pick them (credit preserved):

1. **`HERMES_BASE_HOME` wrongly disabled isolation.** That var "normally points at the base home already" in ordinary deployments, so using it as an opt-out silently turned isolation off for a legitimately-isolated `profiles/<name>`. Fixed by deriving isolation only from the `HERMES_HOME` path shape (also keeps us off a new `HERMES_*` config var per the core rubric).
2. **Isolated profile named literally `default` collapsed explicit resolution to the base home.** `get_hermes_home_for_profile("default")` / `_resolve_profile_home_for_name("default")` hit the root-alias branch → streaming read the wrong config/env/state. Fixed by resolving the isolated profile name (incl. `default`) to `_INITIAL_HERMES_HOME` before the alias branch.
3. **`/api/profiles` reported the upstream base row** for an isolated `profiles/default`. Fixed by only using the upstream row when its resolved path `==` the isolated home, else building from the snapshot (+ fallback `is_default`).

A re-gate on those fixes then surfaced **three deeper items that need your hand** — these are the blockers before this can ship:

**A. (CORE) The aggregate-session path bypasses the fence.** `/api/sessions?all_profiles=1` (and `/api/projects`, `/api/sessions/search`) still return every profile's sessions even when pinned, and the UI can surface this via `_otherProfileCount`. Fix: when `_is_isolated_profile_mode()`, force `all_profiles = False` in those endpoints and return `other_profile_count: 0`.

**B. (CORE — the deep one) `HERMES_HOME=<base>/profiles/default` is not actually isolated from the root/default profile.** WebUI state is scoped by profile *name* (`"default"`), but the WebUI `STATE_DIR` does not follow `HERMES_HOME`, so an isolated `profiles/default` shares sessions/state with root, and `_profiles_match(..., "default")` folds legacy/root rows in. The real fix is to **default `HERMES_WEBUI_STATE_DIR` to `<isolated HERMES_HOME>/webui`** when it's unset and startup `HERMES_HOME` matches `*/profiles/<name>`. ⚠️ This overlaps **#4449** (also about state-dir defaulting) — please reconcile the two so we land one coherent state-dir contract rather than two. Add a `profiles/default` regression proving root/default sessions don't appear.

**C. (SILENT) `switch_profile("default")` in isolated `profiles/default` mode** resolves `home` to `_DEFAULT_HERMES_HOME` instead of the pinned home, so the idempotent same-profile switch can return base config/workspace or mutate process-wide `HERMES_HOME` to base. Fix: resolve the isolated same-name case to `Path(_INITIAL_HERMES_HOME).expanduser()` before the root-alias branch in `switch_profile`.

Take my 3 fixes from the warmup branch as the starting point and add A/B/C on top (B is the one to design carefully + reconcile with #4449). The feature is genuinely close — happy to re-review as soon as the state-dir layer lands. Note this is also a visible Profiles-panel change, so it'll get a screenshot pass before merge.

**Comment by nesquena-hermes** [comment]
Cross-link for the state-dir rework (finding B above): #4449 proposes defaulting the WebUI state dir to `$HERMES_HOME/webui` + a `HERMES_WEBUI_SERVER_CWD` bridge. That's the same contract finding B needs. Please fold #4449's state-dir change into this PR's rework (credit @tomtong2015 via a Co-authored-by trailer) so isolated `profiles/default` gets its own state and the two PRs don't land colliding defaults. #4449 is on hold pending this.

**Comment by rodboev** [comment]
I folded the warmup edge fixes and the `#4449` state-dir contract into this branch instead of splitting the isolation boundary across two overlapping PRs.

1. `api/profiles.py` now treats the startup `HERMES_HOME=.../profiles/<name>` path as authoritative end to end: isolated mode no longer keys off `HERMES_BASE_HOME`, explicit profile-home lookups clamp back to the pinned isolated home, the `profiles/default` row only wins when its resolved path matches the startup home, same-name `default` switches stay pinned, and scheduled cron jobs both execute and publish back under the isolated profile.
2. `api/routes.py` and the profile UI now close the remaining cross-profile affordances: isolated mode returns `single_profile_mode`, hides create/switch/delete/manage controls, ignores `all_profiles=1` on sessions/projects/search, and rejects `all_profiles` CLI imports instead of letting the isolated deployment reach into a foreign profile store through the import path.
3. I folded the `#4449` contract into this PR by defaulting the WebUI state dir to `$HERMES_HOME/webui` when `HERMES_WEBUI_STATE_DIR` is unset in both `api/config.py` and `bootstrap.py`, and added `HERMES_WEBUI_SERVER_CWD` so read-only agent installs can redirect fallback relative writes to a writable workspace on both POSIX and Windows foreground launches.

Focused local validation passed on `tests/test_issue2698_isolated_hermes_home.py tests/test_profile_env_isolation.py tests/test_issue1611_session_profile_filtering.py tests/test_bootstrap_foreground.py tests/test_issue4449_state_dir_contract.py tests/test_issue4067_import_cli_cross_profile_guard.py tests/test_scheduled_jobs_profile_isolation.py tests/test_issue3623_profile_visibility.py tests/test_settings_navigation_and_detail_refresh.py -v --timeout=60` (`119 passed`), and runtime ESLint passed on `static/**/*.js`.

**Comment by nesquena-hermes** [comment]
Shipped in v0.51.528 via the rebased release branch (credited in CHANGELOG + release notes). Thanks @rodboev — strong convergence: the re-push addressed all three deeper findings from the bounce (aggregate-session fence, STATE_DIR-follows-HERMES_HOME, isolated-default resolution). On the way in I added: the README upgrade-migration note + backward-compat state-dir tests (Opus-requested), and fixed two test-infra issues so the suite is green both locally and on CI — an UnboundLocalError from a redundant local import (F823), a suite-ordering leak from reload-based config tests (now subprocess-isolated), and a CI-only ModuleNotFoundError where a test mock.patch'd hermes_cli directly (now stubbed via sys.modules at the WebUI boundary). Full Codex+Opus+suite gate green (9683 local + CI CLEAN). The #4449 state-dir contract is folded in here — credit @tomtong2015. Isolated single-profile mode: point HERMES_HOME at ~/.hermes/profiles/<name>.
